resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -846,22 +861,24 @@ private void tryToFinishResolve(final DnsServerAddressStream nameServerAddrStrea
     }
 
     private void finishResolve(Promise<List<T>> promise, Throwable cause) {
-        if (!queriesInProgress.isEmpty()) {
+        // If completeEarly was true we still want to continue processing the queries to ensure we still put everything
+        // in the cache eventually.
+        if (!completeEarly && !queriesInProgress.isEmpty()) {
             // If there are queries in progress, we should cancel it because we already finished the resolution.
             for (Iterator<Future<AddressedEnvelope<DnsResponse, InetSocketAddress>>> i = queriesInProgress.iterator();
                  i.hasNext();) {
                 Future<AddressedEnvelope<DnsResponse, InetSocketAddress>> f = i.next();
                 i.remove();
 
-                if (!f.cancel(false)) {
-                    f.addListener(RELEASE_RESPONSE);
-                }
+                f.cancel(false);
```

Comment:
please note that I removed the release listener as it will be released anyway in the other listener that is already added to the future when the query is added.